### ISSUE TITLE
To make java-support more friendly to CDI environments

### ISSUE THREAD

**Issue by sleipnir** [cloudstateio/cloudstate#219]
To make java-support more friendly to CDI environments like creating a registerEventSourcedEntity and registerCrdtEntity that accepts a class instance instead of Class.

For that I believe that small changes are needed in Cloudstate.java, AnnotationBaseadEventSourcedSupporte.scala, and AnnotationBasedCrdtSupport.scala.

For more information see: https://github.com/sleipnir/spring-boot-cloudstate-starter/issues/2

**Comment by sleipnir** [comment]
More details: https://github.com/sleipnir/spring-boot-cloudstate-starter/issues/2#issuecomment-604468010

**Comment by sleipnir** [comment]
ping @viktorklang @marcellanz

**Comment by marcellanz** [comment]
@sleipnir sorry to not write yet; too much is happening (at the same time) at the moment.

**Comment by viktorklang** [comment]
@sleipnir Apologies, it's been crazy busy this week. Revisit early next? /cc @jroper

**Comment by sleipnir** [comment]
No problem guys, I’m very busy myself with everything that’s going on too. I know it's hard.

**Comment by sleipnir** [comment]
Looking at the code again, I saw that we only need a new EventSourcedEntityFactory and CrdtEntityFactory that accept a factory class that works as an 'extension' so that each time the create method is called, this fabric delegates to the extension so that we can obtain the instance directly from Spring (in this case the bean must be declared with the prototype Scope so that a new clean object instance is always created), or another context injection.
With that I could use the existing method in Cloudstate.java:

```
public CloudState registerEventSourcedEntity (
      EventSourcedEntityFactory factory,
      Descriptors.ServiceDescriptor descriptor,
      String persistenceId,
      int snapshotEvery,
      Descriptors.FileDescriptor ... additionalDescriptors) {
      ......
}
```

To register user entity classes in a simple way.

It is clear that this new factory should be able to inject the entityId and the Context in a way other than via the constructor. Probably via setter or class attribute.

The question is whether this should be inside java-support or outside. If the class were not really complex as it would be, it would be simple to implement it in the scope of Spring, but as it has a lot of knowledge of the internal structures it may not be so simple to replicate that outside the library......

**Comment by viktorklang** [comment]
@sleipnir TBH It hink having a separate `java-spring-support` module makes more sense. I assume that there'll be others.

**Comment by sleipnir** [comment]
Agree. Is the same is the case with the AKKA Extensions module

**Comment by sleipnir** [comment]
But can having a spring-support and a spring-boot-support make sense? 
Or include this within the module I’m already doing?
Perhaps the module would not be java-spring-support but rather a cloudstate-extensions that would make it easier to run cloudstate in any DI container

**Comment by viktorklang** [comment]
@sleipnir I have no idea—what is the difference—and when would oyu use one over the other?

**Comment by sleipnir** [comment]
@viktorklang The difference is that "Spring" is just the injection container and some modules, as support for aspect-oriented programming, jpa and etc ...
While Spring boot is a Spring project that came to facilitate the process of configuring and publishing our applications. You choose the modules you want through the starters you include in your project's pom.xml. They are, basically, dependencies that group other dependencies.

Despite Spring Boot, through the convention, leaving everything configured, nothing prevents you from creating your customizations if they are necessary. It's ideal for pulling up new microservices and easy to deploy.

The problem related to Cloudstate is related to the ability to inject the dependencies created in the injection container into a class that is not managed by the injection container. So we need a glue to make these pieces come together. Looking at this aspect this is a problem about creating objects and therefore it would make sense for a java-spring module but in production environments applications that use only Spring instead of applications using Spring boot have fallen into disuse. So I created a module for Springboot because it makes more sense in the long run.

But then my provocation in orienting the problem to the context of execution in dependency injection containers because, in this way many other frameworks that have popularity such as Quarkus, spring boot, Dropwizard, or even standalone applications using Weld CDI or Google Guava could be benefited for this approach.

I don't know if I managed to be clear, but I tried to give an overview of the problem

**Comment by viktorklang** [comment]
@sleipnir It would seem to me like it makes most sense to have Boot be supported, because I assume that people will create services from scratch-which means no need to really interop with legacy codebases?

**Comment by sleipnir** [comment]
Perfect. I think the same,

**Comment by sleipnir** [comment]
@viktorklang Maybe a little code can help you understand, I am preparing a simple PR to demonstrate the functionality.

**Comment by viktorklang** [comment]
@sleipnir Cool, I'll have a look. Pinging @jroper on this one too, as DI is one of his favorite things in the world.

**Comment by sleipnir** [comment]
That Cool.
My intention in this PR was not technical preciosity so I'm sorry for the technical flaws, I'm also not the best Dev Scala, I joked a little when the language came up years ago but then I abandoned it because of other language flavors like Elixir and Rust

Thanks @viktorklang @jroper

**Comment by ralphlaude** [comment]
@sleipnir great work!

**Comment by sleipnir** [comment]
Thanks @ralphlaude 
@viktorklang @jroper @ralphlaude I added new documentations in the Springboot client side.
This should help you to have a vision of what the user interface would look like on the Spring side.
I believe that the Readme has enough information to start with and there is a folder of examples. Follow the link:
https://github.com/sleipnir/spring-boot-cloudstate-starter/tree/wip-di-support

**Comment by pvlugter** [comment]
Closed by #282
